spec/app/ics-interchain-queries/README.md

```diff
@@ -0,0 +1,323 @@
+
+```typescript
+function generateQueryIdentifier = () -> Identifier
+```
+
+### Sub-protocols
+
+#### Query lifecycle
+
+1) When the Querying Chain receives a query request, it calls `CrossChainQueryRequest` of the Cross-chain Querying module. This function generates a unique identifier for the query, stores it in its `privateStore` and emits a `sendQuery` event. Query requests can be submitted by other IBC modules as transactions to the Querying Chain or simply executed as part of the `BeginBlock` and `EndBlock` logic.
+2) A correct relayer listening to `sendQuery` events from the Querying Chain will eventually pick the query request up and execute it at the Queried Chain. The result is then submitted (on-chain) to the Querying Chain.
+3) When the query result is committed at the Querying Chain, this calls the `CrossChainQueryResult` function of the Cross-chain Querying module.
+4) The `CrossChainQueryResult` first retrieves the query from the `privateStore` using the query's unique identifier. It then proceeds to verify the result using its local client. If it passes the verification, the function removes the query from the `privateStore` and stores the result in a public path.
+5) The query caller can then asynchronously retrieve the query result.
+
+#### Normal path methods
+
+The `CrossChainQueryRequest` function is called when the Cross-chain Querying module at the Querying Chain receives a new query request.
+
+```typescript
+function CrossChainQueryRequest(
+  path: CommitmentPath,
+  queryHeigth: Heigth,
+  timeoutHeigth: Height,
+  clientId: Identifier,
+  bounty: sdk.Coin,
+  ): Identifier {
+
+    // Check that there exists a client of the Queried Chain. The client will be used to verify the query result.
+    abortTransactionUnless(queryClientState(clientId) !== null)
+
+    // Generate a unique query identifier.
+    queryIdentifier = generateQueryIdentifier()
+
+    // Create a query request record.
+    query = CrossChainQuery{queryIdentifier,
+                            path,
+                            queryHeigth,
+                            timeoutHeigth, 
+                            clientId,
+                            bounty}
+
+    // Store the query in the local, private store.
+    privateStore.set(ongoingQueryPath(queryIdentifier), query)
+
+    // Log the query request.
+    emitLogEntry("sendQuery", query)
+
+    // Returns the query identifier.
+    return queryIdentifier
+}
+```
+- **Precondition**
+  - There exists a client with `clientId` identifier.
+- **Postcondition**
+  - The query request is stored in the `privateStore`.
+  - A `sendQuery` event is emitted.
+
+The `CrossChainQueryResult` function is called when the Cross-chain Querying module at the Querying Chain receives a new query reply.
+
+```typescript
+function CrossChainQueryResult(
+  queryId: Identifier,
+  data: []byte
+  proof: CommitmentProof,
+  proofHeight: Height,
+  success: boolean,
```

Comment:
can't we simplify this by saying that we will verify nonMembership if data is nil?

Seems like the success field is unnecessary

Reply:
Agree, fixed.